src/ir/stack-utils.h

```diff
@@ -158,6 +106,72 @@ struct StackSignature {
     return params == other.params && results == other.results &&
            unreachable == other.unreachable;
   }
+
+  // Whether a block whose contents have stack signature `a` could be typed with
+  // stack signature `b`, i.e. whether it could be used in a context that
+  // expects signature `b`. Formally, where `a` is the LHS and `b` the RHS:
+  //
+  // [t1*] -> [t2*] <: [s1* t1'*] -> [s2* t2'*] iff
+  //
+  //  - t1' <: t1
+  //  - t2 <: t2'
+  //  - s1 <: s2
```

Comment:
Why is the `'` in the LHS in the first of the three clauses? That seems to allow `[anyref] -> [] <: [funcref] -> []` but the LHS cannot be used in a context that expects the RHS?

Reply:
`[anyref] -> []` can be used anywhere `[funcref] -> []` is expected because `funcref` can be consumed by anything that expects an `anyref`. This is contravariance explained in one of the examples below.

Reply:
I apologize, my confusion is almost certainly a matter of language. But I still don't get it:

> [anyref] -> [] can be used anywhere [funcref] -> [] is expected because funcref can be consumed by anything that expects an anyref.

Your statement here uses "expected" in two places. First you say `[funcref] -> [] is expected`, and then you say `expects an anyref`. To me it seems like you're using the term "expected" in two different (reversed) ways, as they are about the exact opposite things, or that you have a typo - ?

Reply:
No worries, I find that I basically have to turn by brain inside out to wrap it around contravariance, so its no wonder everything looks reversed :)

Consider a context that expects an instruction sequence with type `[funcref] -> []`. It looks like this:

```
<prior context>

;; [t* funcref] on stack

[insert our block here]

;; [t*] on stack

<later context>
```

The block we insert needs to consume a `funcref` and produce no values on the stack. What if we actually plug in a block with type `[anyref] -> []`?

```
<prior context>

;; [t* funcref] on stack

[consume `anyref`, produce nothing]

;; [t*] on stack

<later context>
```

Our block expects an `anyref`, but it was provided a `funcref` by the context. This is fine, though, because `funcref` can be used by anything that expects an `anyref`.

The block of type `[anyref] -> []` is valid to use in a context that expects a block of type `[funcref] -> []`, so the former type is a subtype of the latter type.

Reply:
The analogy I found useful to understand is, basically, if `A <: B`, A can act like it is of type B, i.e., Cat can claim it is an Animal and be used it in place anywhere an Animal is expected.

When `A =  [anyref] -> [funcref]` and `B = [funcref] -> [anyref]`, even though `A` takes `anyref`, it can act like it takes `funcref`, because it can process all `funcref` arguments with no problem. (It is called contravariance in parameter types)
Also it can act like it returns `anyref` instead of `funcref`, because all values it returns will fit the condition of `anyref`. (It is called covariance in result types)
So here `A` is a subtype of `B`.

Reply:
I see, thanks! So in `[x] -> [y]` the relation is reversed depending on if the value are consumed or produced... which does in fact make total sense. And then so does the rest.